datafusion/core/src/execution/context.rs

```diff
@@ -1377,6 +1378,8 @@ impl SessionState {
         }
         rules.push(Arc::new(ReduceOuterJoin::new()));
         rules.push(Arc::new(FilterPushDown::new()));
+        // we do type coercion after filter push down so that we don't push CAST filters to Parquet
```

Comment:
I am confused about this comment and explain why do the type coercion after the filter push down optimizer rule.

I think the type coercion rule should be done in preview stage.

Reply:
For example,  `Filter expr: FLOAT32(C1) < FLOAT64(16)`. We should do type coercion first and convert the filter expr to `CAST(INT32(C1) AS FLOAT64 < FLOAT64(16)` and try to push the new filter expr to the table scan operation.

If you don't do the type coercion first, you will push the `expr: FLOAT32(C1) < FLOAT64(16)` to table scan, Does this can be applied to the parquet filter or pruning filter?

Reply:
> @liukun4515 This PR is ready for review now

Yes, this is ready for review now.

Reply:
I filed https://github.com/apache/arrow-datafusion/issues/3289 applying `TypeCoercion` before `FilterPushDown`. I think the PR would get too large to review if I make those changes here.